lib/byron/src/Cardano/Wallet/Byron/Network.hs

```diff
@@ -739,6 +893,105 @@ localTxSubmission tr queue channel =
+--        │       ┌───────────┴───┐             ┌────────┴───────┐
+--        └───────┤   Acquired    │────────────>│   Querying     │
+--                └───────────────┘             └────────────────┘
+--
+localStateQuery
+    :: forall m protocol.
+        ( MonadThrow m, MonadTimer m, MonadST m
+        , protocol ~ LocalStateQuery ByronBlock (Query ByronBlock)
+        )
+    => Tracer m (TraceSendRecv protocol)
+        -- ^ Base tracer for the mini-protocols
+    -> TQueue m (LocalStateQueryCmd m)
+        -- ^ We use a 'TQueue' as a communication channel to drive queries from
+        -- outside of the network client to the client itself.
+        -- Requests are pushed to the queue which are then transformed into
+        -- messages to keep the state-machine moving.
+    -> Channel m ByteString
+        -- ^ A 'Channel' is a abstract communication instrument which
+        -- transports serialized messages between peers (e.g. a unix
+        -- socket).
+    -> m Void
+localStateQuery tr queue channel =
+    runPeer tr codec channel (localStateQueryClientPeer client)
+  where
+    -- fixme: Marcin says use defaultCodecs or clientCodecs instead of this
+    codec :: Codec protocol DeserialiseFailure m ByteString
+    codec = codecLocalStateQuery
+          (encodePoint encodeByronHeaderHash)
+          (decodePoint decodeByronHeaderHash)
+          nodeEncodeQuery
+          nodeDecodeQuery
+          nodeEncodeResult
+          nodeDecodeResult
+
+    client
+        :: LocalStateQueryClient ByronBlock (Query ByronBlock) m Void
+    client = LocalStateQueryClient clientStIdle
+      where
+        clientStIdle
+            :: m (LocalStateQuery.ClientStIdle ByronBlock (Query ByronBlock) m Void)
+        clientStIdle = atomically (readTQueue queue) <&> \case
+            CmdQueryLocalState pt respond ->
+                LocalStateQuery.SendMsgAcquire pt (clientStAcquiring respond)
+
+        clientStAcquiring
+            :: (LocalStateQueryResult -> m ())
+            -> LocalStateQuery.ClientStAcquiring ByronBlock (Query ByronBlock) m Void
+        clientStAcquiring respond = LocalStateQuery.ClientStAcquiring
+            { recvMsgAcquired = clientStAcquired1 respond
+            , recvMsgFailure = \failure -> do
+                    respond (Left failure)
+                    clientStIdle
+            }
+
+        clientStAcquired1
+            :: (LocalStateQueryResult -> m ())
+            -> LocalStateQuery.ClientStAcquired ByronBlock (Query ByronBlock) m Void
+        clientStAcquired1 respond = LocalStateQuery.SendMsgQuery GetUpdateInterfaceState (clientStQuerying respond)
+
+        clientStAcquired2
+            :: m (LocalStateQuery.ClientStAcquired ByronBlock (Query ByronBlock) m Void)
+        clientStAcquired2 = atomically (readTQueue queue) <&> \case
+            CmdQueryLocalState pt respond ->
+                LocalStateQuery.SendMsgReAcquire pt (clientStAcquiring respond)
```

Comment:
Is this (re-acquiring) preferred over releasing the state and acquiring it back later ?

Reply:
Yes, I added a comment.